kernel/uefi.c

```diff
@@ -0,0 +1,159 @@
+#include <tilck/common/string_util.h>
+#include <tilck/common/printk.h>
+
+#include <tilck/kernel/kmalloc.h>
+#include <tilck/kernel/system_mmap.h>
+#include <tilck/kernel/process.h>
+
+#define EFI_DEBUG 0
+#include <efi.h>
+#include <efiapi.h>
+
+ulong uefi_rt_addr;
+
+void setup_uefi_runtime_services(void)
+{
+   struct mem_region ma;
+   EFI_RUNTIME_SERVICES *RT;
+   EFI_MEMORY_DESCRIPTOR *virt_map;
+
+   int num_entries = 0;
+
+   if (!uefi_rt_addr)
+      return;
+
+   RT = TO_PTR(uefi_rt_addr);
+   printk("Runtime Services: %p\n", RT);
+
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+
+      get_mem_region(i, &ma);
+
+      if (ma.type == TILCK_BOOT_EFI_RUNTIME_CODE ||
+          ma.type == TILCK_BOOT_EFI_RUNTIME_DATA)
+      {
+         num_entries++;
+      }
+   }
+
+   /*
+    * When SetVirtualAddressMap() is called, the physical
+    * addresses should be directly accessible. So, identity
+    * map them.
+    */
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+
+      get_mem_region(i, &ma);
+
+      if (ma.type != TILCK_BOOT_EFI_RUNTIME_CODE &&
+          ma.type != TILCK_BOOT_EFI_RUNTIME_DATA)
+      {
+         continue;
+      }
+
+      const ulong pbegin = (ulong)ma.addr;
+      const ulong pend = MIN((ulong)(ma.addr+ma.len), (ulong)LINEAR_MAPPING_SIZE);
+
+      const size_t page_count = (pend - pbegin) >> PAGE_SHIFT;
+
+      ulong *vbegin = TO_PTR(pbegin);
+
+      /*
+       * SetupVirtualAddressMap() requires the runtime data
+       * as well as runtime code to be writable. So use
+       * `PAGING_FL_RW` for both.
+       */
+      size_t count =
+         map_pages(get_kernel_pdir(),
+                   vbegin,
+                   pbegin,
+                   page_count,
+                   PAGING_FL_RW);
+
+      if (count != page_count)
+         panic("kmalloc: unable to map regions in the virtual space");
+
+      num_entries++;
+   }
+
+   virt_map = kmalloc((size_t)num_entries);
+   num_entries = 0;
+
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+
+      EFI_MEMORY_DESCRIPTOR desc;
+
+      get_mem_region(i, &ma);
+
+      if (ma.type != TILCK_BOOT_EFI_RUNTIME_CODE &&
+          ma.type != TILCK_BOOT_EFI_RUNTIME_DATA)
+      {
+         continue;
+      }
+
+      desc.Type =
+         ma.type == TILCK_BOOT_EFI_RUNTIME_CODE
+            ? EfiRuntimeServicesCode
+            : EfiRuntimeServicesData;
+
+      desc.PhysicalStart = (ulong)ma.addr;
+      desc.VirtualStart = (ulong)KERNEL_PA_TO_VA(ma.addr);
+      desc.NumberOfPages = ma.len / PAGE_SIZE;
+      desc.Attribute = EFI_MEMORY_RUNTIME;
+
+      virt_map[num_entries++] = desc;
+
+      printk("Type: %d\tPhysical Address: %lld\t Virtual Address: %lld\n", desc.Type, desc.PhysicalStart, desc.VirtualStart);
+   }
+
+   printk("Virtual map: %p\n", virt_map);
+
+   printk("Time %p\n", RT->GetTime);
+   printk("SVAM %p\n", RT->SetVirtualAddressMap);
+
+   UINTN mem_map_size = num_entries * sizeof(EFI_MEMORY_DESCRIPTOR);
```

Comment:
Can't we skip all of those extra variables? Just declare `EFI_STATUS status` above and simply do:
```
status = RT->SetVirtualAddressMap(num_entries * sizeof(EFI_MEMORY_DESCRIPTOR),
                                  sizeof(EFI_MEMORY_DESCRIPTOR), 
                                  EFI_MEMORY_DESCRIPTOR_VERSION,
                                  virt_map);
```

After that, checking that `status` is a success is mandatory. We can panic in case of a failure.